sahara.c

```diff
@@ -246,14 +301,149 @@ static int sahara_done(struct qdl_device *qdl, struct sahara_pkt *pkt)
 	return pkt->done_resp.status;
 }
 
-int sahara_run(struct qdl_device *qdl, char *img_arr[], bool single_image)
+static ssize_t sahara_debug64_one(struct qdl_device *qdl,
+				  struct sahara_debug_region64 region,
+				  int ramdump_dir)
+{
+	struct sahara_pkt read_req;
+	uint64_t remain;
+	size_t offset;
+	size_t chunk;
+	ssize_t n;
+	void *buf;
+	int fd;
+
+	buf = malloc(DEBUG_BLOCK_SIZE);
+	if (!buf)
+		return -1;
+
+	fd = openat(ramdump_dir, region.filename, O_WRONLY | O_CREAT, 0644);
+	if (fd < 0) {
+		warn("failed to open \"%s\"", region.filename);
+		return -1;
+	}
+
+	chunk = 0;
+	while (chunk < region.length) {
+		remain = MIN(region.length - chunk, DEBUG_BLOCK_SIZE);
+
+		read_req.cmd = SAHARA_MEM_READ64_CMD;
+		read_req.length = SAHARA_MEM_READ64_LENGTH;
+		read_req.debug64_req.addr = region.addr + chunk;
+		read_req.debug64_req.length = remain;
+		n = qdl_write(qdl, &read_req, read_req.length);
+		if (n < 0)
+			break;
+
+		offset = 0;
+		while (offset < remain) {
+			n = qdl_read(qdl, buf, DEBUG_BLOCK_SIZE, 30000);
+			if (n < 0) {
+				warn("failed to read ramdump chunk");
+				goto out;
+			}
+
+			write(fd, buf, n);
```

Comment:
I built this real quick today (in hopes of testing it, but that's looking like another day now lol), and noticed this:

```
# make
cc -O2 -Wall -g `pkg-config --cflags libxml-2.0`   -c -o qdl.o qdl.c
cc -O2 -Wall -g `pkg-config --cflags libxml-2.0`   -c -o sahara.o sahara.c
sahara.c: In function ‘sahara_debug64_one’:
sahara.c:346:4: warning: ignoring return value of ‘write’, declared with attribute warn_unused_result [-Wunused-result]
    write(fd, buf, n);
    ^~~~~~~~~~~~~~~~~
```